spot-ml/src/main/scala/org/apache/spot/dns/DNSSuspiciousConnectsAnalysis.scala

```diff
@@ -48,34 +27,84 @@ object DNSSuspiciousConnectsAnalysis {
+          inputDataFrame: DataFrame) = {
 
-    logger.info("Loading data")
+    logger.info("Starting DNS suspicious connects analysis.")
 
     val userDomain = config.userDomain
 
-    val rawDataDF = sqlContext.read.parquet(config.inputPath)
-      .filter(Timestamp + " is not null and " + UnixTimestamp + " is not null")
-      .select(inColumns:_*)
+    val cleanDataDF = inputDataFrame
+      .filter(InputFilter)
+      .select(InSchema:_*)
+      .na.fill("unknown", Seq(QueryClass))
+      .na.fill(-1, Seq(QueryType))
+      .na.fill(-1, Seq(QueryResponseCode))
 
     logger.info("Training the model")
 
     val model =
-      DNSSuspiciousConnectsModel.trainNewModel(sparkContext, sqlContext, logger, config, rawDataDF, config.topicCount)
+      DNSSuspiciousConnectsModel.trainNewModel(sparkContext, sqlContext, logger, config, cleanDataDF, config.topicCount)
 
     logger.info("Scoring")
-    val scoredDF = model.score(sparkContext, sqlContext, rawDataDF, userDomain)
+    val scoredDF = model.score(sparkContext, sqlContext, cleanDataDF, userDomain)
+
+    val filteredDF = scoredDF.filter(Score + " <= " + config.threshold + " AND " + Score + " > -1")
 
+    val orderedDF = filteredDF.orderBy(Score)
 
-    val filteredDF = scoredDF.filter(Score + " <= " + config.threshold)
-    val mostSusipiciousDF: DataFrame = filteredDF.orderBy(Score).limit(config.maxResults)
+    val mostSuspiciousDF = if(config.maxResults > 0)  orderedDF.limit(config.maxResults) else orderedDF
 
-    val outputDF = mostSusipiciousDF.select(OutColumns:_*).sort(Score)
+    val outputDF = mostSuspiciousDF.select(OutSchema:_*).sort(Score)
 
-    logger.info("DNS  suspcicious connects analysis completed.")
+    logger.info("DNS  suspicious connects analysis completed.")
     logger.info("Saving results to : " + config.hdfsScoredConnect)
     outputDF.map(_.mkString(config.outputDelimiter)).saveAsTextFile(config.hdfsScoredConnect)
+
+    val invalidRecords = inputDataFrame
+      .filter(InvalidRecordsFilter)
+      .select(InSchema:_*)
+
+    dataValidation.showAndSaveInvalidRecords(invalidRecords, config.hdfsScoredConnect, logger)
+
+    val corruptRecords = scoredDF.filter(Score + " = -1")
+    dataValidation.showAndSaveCorruptRecords(corruptRecords, config.hdfsScoredConnect, logger)
   }
+
+
+  val InStructType = StructType(List(TimestampField, UnixTimestampField, FrameLengthField, ClientIPField,
+    QueryNameField, QueryClassField, QueryTypeField, QueryResponseCodeField))
+
+  val InSchema = InStructType.fieldNames.map(col)
+
+  assert(ModelSchema.fields.forall(InStructType.fields.contains(_)))
+
+  val OutSchema = StructType(
+    List(TimestampField,
+      UnixTimestampField,
+      FrameLengthField,
+      ClientIPField,
+      QueryNameField,
+      QueryClassField,
+      QueryTypeField,
+      QueryResponseCodeField,
+      ScoreField)).fieldNames.map(col)
+
+  val InputFilter = s"($Timestamp IS NOT NULL AND $Timestamp <> '' AND $Timestamp <> '-') " +
```

Comment:
I generally don't like SQL strings in code.... bugs show up only at run time
but it is close to the domain area and easy to read
so i'm cool with it if it gets unit tested